### PR TITLE
Add nanobind 2.2.0 and 2.4.0

### PR DESCRIPTION
**Pull request by keith**
This also makes a few minor changes to the BUILD file:

- This removes the global `-undefined dynamic_lookup` on macOS and uses
  the provided symbols file instead to only ignore symbols that nanobind
  knows will come from python
- This removes some compiler flags that I don't think nanobind should be
  deciding for the project, specifically `-Os` and `-flto`, IMO these
  should be configured by the user if desired
- This uses the current python headers instead of those from python 3.8,
  as depending on the user's project they might not want to use 3.8
  specifically, this can be changed via a transition if needed, but
  otherwise will pull the user's default toolchain
- This adds -frtti which is required similar to exceptions

### CODE REVIEW

**Comment by keith** [comment]
@adzenith ptal and lmk what you think about the BUILD file changes here

**Comment by bazel-io** [comment]
Hello @adzenith, modules you maintain (nanobind) have been updated in this PR. Please review the changes.

File: modules/nanobind/2.2.0/MODULE.bazel [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module(
+    name = "nanobind",
+    version = "2.2.0",
+    compatibility_level = 1,
+)
+
+bazel_dep(name = "platforms", version = "0.0.10")
+bazel_dep(name = "robin-map", version = "1.3.0")
+bazel_dep(name = "rules_python", version = "0.36.0")
```
Comment (adzenith) [inline modules/nanobind/2.2.0/MODULE.bazel:9]: 1.0.0 is out - do we want to upgrade?

File: modules/nanobind/2.4.0/patches/module_dot_bazel.patch [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+@@ -0,0 +1,9 @@
++module(
++    name = "nanobind",
++    version = "2.4.0",
++    compatibility_level = 1,
++)
++
++bazel_dep(name = "platforms", version = "0.0.10")
++bazel_dep(name = "robin-map", version = "1.3.0")
++bazel_dep(name = "rules_python", version = "0.36.0")
```
Comment (adzenith) [inline modules/nanobind/2.4.0/patches/module_dot_bazel.patch:15]: same here

File: modules/nanobind/2.2.0/MODULE.bazel (hunk 1 above)
Comment (keith) [reply modules/nanobind/2.2.0/MODULE.bazel:9]: the downside of doing that is it forces all consumers to use that at a minimum, so it's generally preferred that we use our lowest supported version here, so users can use `>=` this

**Review by adzenith — APPROVED** [review]
Looks great! Thank you!

**Review by bazel-io — APPROVED** [review]
Hello @bazelbuild/bcr-maintainers, all modules in this PR have been approved by their maintainers. Please take a final look to merge this PR.

**Pull request merged**